AndroidHomeApplication/app/src/main/java/com/example/androidhomeapplication/ui/moviesList/FragmentMoviesList.kt

```diff
@@ -26,39 +29,96 @@ class FragmentMoviesList : Fragment(R.layout.fragment_movies_list) {
+    var isLoading = false
+
+    private val adapter: MoviesListAdapter = MoviesListAdapter(
         onItemClick = { item ->
-            (activity?.application as? RouterProvider)?.router?.navigateTo(MovieDetailsScreen(item.id))
+            val cacheResult = viewModel.searchQuery.isEmpty()
+            (activity?.application as? RouterProvider)?.router?.navigateTo(MovieDetailsScreen(item.id, cacheResult))
         })
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
 
         binding.cinemaRv.layoutManager = GridLayoutManager(context, 2)
-        binding.cinemaRv.adapter = pagingAdapter
+        binding.cinemaRv.adapter = adapter
 
         initListeners()
-
     }
 
+
     private fun initListeners() {
 
+        viewModel.viewState.observe(viewLifecycleOwner) { state ->
+            when (state) {
+                is State.Empty -> {
+                    adapter.submitList(emptyList())
+                    viewModel.getCachedMovies()
+                    Log.d("FragmentMoviesList_LOG", "State.Empty")
+                }
+                is State.EmptyProgress -> {
+                    Log.d("FragmentMoviesList_LOG", "State.EmptyProgress")
+                }
+                is State.Data<*> -> {
+                    val movies = state.data as List<Movie>
+                    adapter.submitList(movies)
+
+                    if (viewModel.searchQuery.isEmpty()){
+                        viewModel.saveMoviesToCache(movies)
+                    }
+
+                    Log.d("FragmentMoviesList_LOG", "State.Data: page - ${state.page}")
+                }
+                is State.Refresh<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.Refresh: page - ${state.page}")
+                }
+                is State.NewPageProgress<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.NewPageProgress: page - ${state.page}")
+                }
+                is State.FullData<*> -> {
+                    Log.d("FragmentMoviesList_LOG", "State.FullData: page - ${state.page}")
+                }
+            }
+        }
+
+        viewModel.cachedMovies.observe(viewLifecycleOwner) { movies ->
+            if (viewModel.viewState.value == State.Empty){
+                adapter.submitList(movies)
+                viewModel.refresh()
+                Log.d("FragmentMoviesList_LOG", "Show Cached data and call refresh")
+            }
+        }
+
         binding.textInputLayout.textInputEditText.textChanges()
-            .debounce(500)
+            .debounce(300)
             .map { charSequence -> charSequence?.toString() }
             .distinctUntilChanged()
             .map { value ->
-                viewModel.loadData(value)
-                pagingAdapter.refresh()
+                if (viewModel.searchQuery != value) {
+                    adapter.submitList(emptyList())
```

Comment:
adapter.submitList(emptyList())
should be done as part of changing state via reducer